lib/app/routes/exercises.rb

```diff
@@ -119,6 +119,19 @@ class Exercises < Core
         redirect "/submissions/#{submission.key}"
       end
 
+      post '/submissions/:key/hibernate' do |key|
+        please_login("You have to be logged in to do that")
+        submission = Submission.find_by_key(key)
+        unless current_user.owns?(submission)
+          flash[:notice] = "Only the submitter may hibernate this exercise."
+          redirect "/submissions/#{key}"
+        end
+        submission.state = "hibernating"
```

Comment:
There's a bit of trickiness here, due to some legacy duct-tape.

The first is that if someone chooses "hibernate" on a submission that is not the last one, we'll get an exercise that is in a very weird state, and probably won't be considered hibernated by much of the code.

The other is that there's a sort-of related object called a UserExercise, which has a denormalized state that should match (did I mention duct tape?).

Look for `Hack::UpdatesUserExercise`, e.g.

```
Hack::UpdatesUserExercise.new(submission.user_id, submission.track_id, submission.slug).update
```